datafusion/physical-expr/src/expressions/case.rs

```diff
@@ -635,4 +704,67 @@ mod tests {
             RecordBatch::try_from_iter(vec![("load4", Arc::new(load4) as ArrayRef)])?;
         Ok(batch)
     }
+
+    #[test]
+    fn case_test_incompatible() -> Result<()> {
+        // 1 then is int64
+        // 2 then is boolean
+        let batch = case_test_batch()?;
+        let schema = batch.schema();
+
+        // CASE WHEN a = 'foo' THEN 123 WHEN a = 'bar' THEN true END
+        let when1 = binary(
+            col("a", &schema)?,
+            Operator::Eq,
+            lit(ScalarValue::Utf8(Some("foo".to_string()))),
```

Comment:
It would be nice if the the `lit` in the physical expr could use `lit("foo")` as well. But I tried it and it doesn't work 

```suggestion
            lit("foo"),
```

Reply:
https://github.com/apache/arrow-datafusion/pull/2828

Reply:
I will rebase this pr after https://github.com/apache/arrow-datafusion/pull/2828 merged

Reply:
changed the code to use `lit("foo")` with merged https://github.com/apache/arrow-datafusion/pull/2828